### PR TITLE
Manifest

### LINKED ISSUE #858: Manifest parser should support extensions
Currently the Manifest parser aborts when there's any field unexpected in the Manifest. This is a pain every time we are trying to add a new feature to the Manifest. We could abort only when any supported feature is different from what's expected, but continue normally if there new fields or flags.

### PR DESCRIPTION
**Pull request by otaviobp**
Reorganizing manifest parser, adding some improvements and making it easier to extend Manifests without breaking swupd. 

Fixes #858

### CODE REVIEW

File: src/manifest_parser.c [hunk 1]
```diff
@@ -0,0 +1,313 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2012-2016 Intel Corporation.
```
Comment (castulo) [inline src/manifest_parser.c]: 2016?

File: src/manifest_parser.c [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+	if (strncmp(line, "MANIFEST\t", 9) != 0) {
+		goto err_close;
+	}
+
+	c = &line[9];
+	err = strtoi_err(c, &manifest_enc_version);
+
+	if (manifest_enc_version <= 0 || err != 0) {
+		fprintf(stderr, "Error: Loaded incompatible manifest version\n");
```
Comment (castulo) [inline src/manifest_parser.c]: errors should use the error() function

File: src/manifest_parser.c [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+				 * to be big enough to allow Manifest.Full to
+				 * pass (currently about 450,000 March 18) but
+				 * small enough that when multiplied by
+				 * sizeof(struct file) it fits into
+				 * size_t. For a system with size_t being a 32
+				 * bit value, this constrains it to be less
+				 * than about 6,000,000, but close to infinity
+				 * for systems with 64 bit size_t.
+				 */
+				fprintf(stderr, "Error: preposterous (%llu) number of files in %s Manifest, more than 4 million skipping\n",
```
Comment (castulo) [inline src/manifest_parser.c]: same comment for all error messages in this section, use error() function

File: src/manifest.c [hunk 4]
```diff
@@ -77,323 +77,47 @@ int file_sort_filename_reverse(const void *a, const void *b)
...
-	struct manifest *manifest;
-	struct list *includes = NULL;
 	char *filename;
 	char *basedir;
-	unsigned long long filecount = 0;
-	unsigned long long contentsize = 0;
-	int manifest_hdr_version;
-	int manifest_enc_version;
+	struct manifest *manifest;
+	struct list *l;
```
Comment (castulo) [inline src/manifest.c]: I would prefer if we use variable names that are more descriptive and longer than one letter, that makes the code more readable.

File: src/manifest.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#endif
+
+struct manifest {
+	// Header
+	int manifest_version;
+	int version;
+	uint64_t filecount;
+	uint64_t contentsize;
+	struct list *includes; /* list of strings with manifest names */
+	//struct list *actions;     /* list of strings with actions */
```
Comment (castulo) [inline src/manifest.h]: leftover?

File: src/manifest.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	// File list
+	struct list *files;     /* struct file for files */
+	struct list *manifests; /* struct file for possible manifests */
+
+	// Helper data
+	struct list *submanifests; /* struct manifest for subscribed manifests */
+	unsigned int is_mix : 1;
+};
+
+struct manifest *manifest_parse(const char *component, const char *filename, bool header_only);
```
Comment (castulo) [inline src/manifest.h]: Most other header files include a brief description of what the function is for, this one doesn't.

File: src/manifest.c (hunk 4 above)
Comment (otaviobp) [reply src/manifest.c]: This list is used as an iterator in a for, so I don't think we need a better name, but I could move it to a proper place, just before it's used.

File: src/manifest.h (hunk 6 above)
Comment (otaviobp) [reply src/manifest.h]: haha, I totally forgot to add comments here.

**Comment by otaviobp** [comment]
Fixed the proposed changes

**Pull request merged**